src/drivers/sht3xd.c

```diff
@@ -0,0 +1,287 @@
+/**
+ * @section License
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2017, Google Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person
+ * obtaining a copy of this software and associated documentation
+ * files (the "Software"), to deal in the Software without
+ * restriction, including without limitation the rights to use, copy,
+ * modify, merge, publish, distribute, sublicense, and/or sell copies
+ * of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ * This code was originally written by Simon L. B. Nielsen.
+ * This file is part of the Simba project.
+ */
+
+#include "simba.h"
+
+/*
+ * This driver supports the Sensirion SHT3x-D series of digital of
+ * Humidity and Temperature Sensors, specifically the SHT30-D,
+ * SHT31-D, and SHT35-D.
+ *
+ * Datasheet:
+ * https://www.sensirion.com/fileadmin/user_upload/customers/sensirion/Dokumente/2_Humidity_Sensors/Sensirion_Humidity_Sensors_SHT3x_Datasheet_digital.pdf
+ *
+ * Reference code:
+ * https://www.sensirion.com/fileadmin/user_upload/customers/sensirion/Dokumente/11_Sample_Codes_Software/Humidity_Sensors/Sensirion_Humidity_Sensors_SHT3x_Sample_Code_V2.pdf
+ */
+
+#if CONFIG_SHT3XD == 1
+
+#define CMD_SINGLE_CLKSTRETCH_REPHIGH     0x2c06
+#define CMD_SINGLE_NOCLKSTRETCH_REPHIGH   0x2400
+#define CMD_GETSERIAL                     0x3780
+
+/**
+ * Send a single 16bit command to the SHT3x-D device.
+ */
+static int sht3xd_sendcmd(struct sht3xd_driver_t *self_p, uint16_t cmd)
+{
+    ASSERTN(self_p != NULL, EINVAL);
```

Comment:
Precondition checks in non-exported functions are not required, nor recommended. They increase the binary size, and are unlikely to catch any bugs since they are internal.